### ISSUE TITLE
m3 reporter: May silently drop metrics if the queue is full

### ISSUE THREAD

**Issue by prashantv** [uber-go/tally#153]
There's a TODO in the code for this:
https://github.com/uber-go/tally/blob/master/m3/reporter.go#L478
```
		select {
		case r.metCh <- sm:
		default:
			// TODO: don't drop when full, or add metric to track
		}
```

When a large number of metrics have changed since the last report, it's likely that metrics will be silently dropped. This becomes more likely since we report all changed metrics at once on a timer tick.

It's hard to know the impact of this as it's silent to consumers.

**Comment by mway** [comment]
I added metrics that include this in [f8fad1e](https://github.com/uber-go/tally/commit/f8fad1e376b3f096ae00349a3556b9f045f020b8) (soon to be PR'ed), but I've also been considering changing this to always block. The reality is that, with a scope flush interval of 1s, and with current benchmarks, there's an approximate throughput of well over 1M metrics/s (closer to 2-3M on my machine), which seems more than sufficient (noting that this is at the reporter level, meaning aggregated datapoints for unique timeseries, not individual metric updates).

Unless we can think of a reason not to block, or to make it conditional based on config, I can include this with the upcoming changes.

**Comment by prashantv** [comment]
Internal metrics look great, thanks @mway! The gauges may end up overwriting each other, so histograms may be helpful.

Re: removing the drop completely, agree there we probably shouldn't drop at all.

**Comment by noahgoldman** [comment]
Should this issue now be closed?  Looks like [v3.4.1](https://github.com/uber-go/tally/releases/tag/v3.4.1) includes the fix.

**Comment by prashantv** [comment]
Agree, closing as 3.4.1 includes the fix.
